src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java

```diff
@@ -31,23 +32,24 @@ public Product getProduct(Long id) throws ProductNotFoundException {
     @Override
     synchronized public void save(Product product) {
         if (product.getId() != null) {
-            productList.stream()
+            Optional<Product> foundedProduct = productList.stream()
                     .filter(p -> product.getId().equals(p.getId()))
-                    .findAny()
-                    .ifPresent(p -> p = product);
-        }
-        else {
-            product.setId(maxId++);
-            productList.add(product);
+                    .findAny();
+            if (foundedProduct.isPresent()) {
+                foundedProduct.get().replaceProduct(product);
```

Comment:
Please, move replaceProduct(product) method from Product class to ArrayListProductDao. It's a bad practice placing logic in domain layer classes